clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -347,9 +425,157 @@ event_inscount_thread_init(void *drcontext)
+              path_to_file.c_str());
+    }
+
+    // Enforcing constraints on intervals:
+    // 1) They need to be ordered by start time.
+    std::sort(instr_intervals.begin(), instr_intervals.end(), cmp_start_instr);
+
+    // 2) Overlapping intervals must be merged.
+    std::vector<instr_interval_t> instr_intervals_merged;
+    instr_intervals_merged.emplace_back(instr_intervals[0]);
+    for (instr_interval_t &interval : instr_intervals) {
+        uint64 end = interval.start + interval.duration;
+        instr_interval_t &last_interval = instr_intervals_merged.back();
+        uint64 last_end = last_interval.start + last_interval.duration;
+        if (interval.start <= last_end) {
+            uint64 max_end = last_end > end ? last_end : end;
+            last_interval.duration = max_end - last_interval.start;
+        } else {
+            instr_intervals_merged.emplace_back(interval);
+        }
+    }
+
+    return instr_intervals_merged;
+}
+
+static void
+free_trace_window_entry(void *entry)
+{
+    dr_global_free(entry, sizeof(uint64));
+}
+
+// Transforms instruction intervals from <start,duration> pairs to trace and no_trace
+// number of instructions. Has the side effect of populating the read-only, global vectors
+// trace_for_instrs_list and no_trace_for_instrs_list, and the global num_windows.
+static void
+compute_irregular_trace_windows(std::vector<instr_interval_t> &instr_intervals)
+{
+    if (instr_intervals.empty())
+        return;
+
+    uint num_intervals = (uint)instr_intervals.size();
+    num_windows = num_intervals + 1;
+
+    drvector_init(&trace_for_instrs_list, num_windows, false, free_trace_window_entry);
+    drvector_init(&no_trace_for_instrs_list, num_windows, false, free_trace_window_entry);
+
+    uint64 *trace_for_instrs_ptr = (uint64 *)dr_global_alloc(sizeof(uint64));
+    *trace_for_instrs_ptr = instr_intervals[0].duration;
+    drvector_set_entry(&trace_for_instrs_list, 0, trace_for_instrs_ptr);
+
+    uint64 *no_trace_for_instrs_ptr = (uint64 *)dr_global_alloc(sizeof(uint64));
+    *no_trace_for_instrs_ptr = instr_intervals[0].start;
+    drvector_set_entry(&no_trace_for_instrs_list, 0, no_trace_for_instrs_ptr);
+
+    for (uint i = 1; i < num_intervals; ++i) {
+        uint64 trace_for_instrs = instr_intervals[i].duration;
+        uint64 no_trace_for_instrs = instr_intervals[i].start -
+            (instr_intervals[i - 1].start + instr_intervals[i - 1].duration);
+
+        trace_for_instrs_ptr = (uint64 *)dr_global_alloc(sizeof(uint64));
+        *trace_for_instrs_ptr = trace_for_instrs;
+        drvector_set_entry(&trace_for_instrs_list, i, trace_for_instrs_ptr);
+
+        no_trace_for_instrs_ptr = (uint64 *)dr_global_alloc(sizeof(uint64));
+        *no_trace_for_instrs_ptr = no_trace_for_instrs;
+        drvector_set_entry(&no_trace_for_instrs_list, i, no_trace_for_instrs_ptr);
+    }
+
+    trace_for_instrs_ptr = (uint64 *)dr_global_alloc(sizeof(uint64));
+    *trace_for_instrs_ptr = 0;
```

Comment:
The assumption when using irregular windows is that the user knows (more or less) how many instructions the target program executes (otherwise they wouldn't know when to start the instruction intervals or how long they should be). If they want to end with trace-forever they can make the last window large enough.